eni: sync all nodes on operator starting

wu0407 opened the pull request:
Set all nodes on the field `NodeManager.nodes` before `NodeManager.Start` , that
fix `NodeManager.Resync` not resync immediately all nodes  reconciling the IP resource pool  and metrics `cilium_operator_ipam_nodes` and
`cilium_operator_ipam_available_interfaces` not corrected.


Signed-off-by: xiaoqing <xiaoqingnb@gmail.com>

Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

<!-- Description of change -->

Fixes: #20815

```release-note
sync all nodes on operator starting
```

gandro reviewed (changes requested):
Thanks. If I understand the proposal correctly, this is to fix an issue with the metrics not having seen all nodes yet.

The way we typically solve this problem is by watching all nodes and waiting for the k8s client to confirm that we have seen all  (by waiting for `WaitForCacheSync`) instead of listing all nodes explicitly, which then calls `NodeManager.Resync`. We're not doing this for ENI mode at the moment, but we could absolutely add this.

https://github.com/cilium/cilium/blob/e0537f91dee911a36044c9a21fea0e875935a38e/operator/cmd/root.go#L546

https://github.com/cilium/cilium/blob/344863ac55e30f88706db06aca5af9d79db4c00a/pkg/ipam/node_manager.go#L458

`Resync` would then be called once it is guaranteed that the `NodeManager` has seen every node. Why is it not possible to initialize the metrics there?

christarazi reviewed (changes requested):
Echoing Sebastian's comment as well. Why can't we rely on the `Update()` call to the Node Manager from K8s when a new node is created to execute the code in this PR?

wu0407 committed:
eni: sync all nodes on operator starting

Set all nodes on the field `NodeManager.nodes` before `NodeManager.Start`, that
fix `NodeManager.Resync` not resync immediately all nodes to resolves IP
deficits and metrics `cilium_operator_ipam_nodes` and
`cilium_operator_ipam_available_interfaces` not corrected.

Fixes: #20815

Signed-off-by: xiaoqing <xiaoqingnb@gmail.com>
